linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -2897,16 +2990,86 @@ private[optimizer] abstract class OptimizerCore(
+              BinaryOp.Long_==, LongLiteral(0L))
+        } else {
+          pretransformApply(ApplyFlags.empty, targs.head,
+              MethodIdent(LongImpl.divideUnsigned), targs.tail,
+              ClassType(LongImpl.RuntimeLongClass), isStat, usePreTransform)(
+              cont)
+        }
       case LongRemainderUnsigned =>
-        pretransformApply(ApplyFlags.empty, targs.head,
-            MethodIdent(LongImpl.remainderUnsigned), targs.tail,
-            ClassType(LongImpl.RuntimeLongClass), isStat, usePreTransform)(
-            cont)
+        if (isWasm) {
+          genericWasmDivModUnsigned(WasmBinaryOp.I64RemU, BinaryOp.Long_%,
+              BinaryOp.Long_==, LongLiteral(0L))
+        } else {
+          pretransformApply(ApplyFlags.empty, targs.head,
+              MethodIdent(LongImpl.remainderUnsigned), targs.tail,
+              ClassType(LongImpl.RuntimeLongClass), isStat, usePreTransform)(
+              cont)
+        }
+
+      // java.lang.Float
+
+      case FloatToIntBits =>
+        // The Wasm I32ReinterpretF32 is the *raw* version; we need to normalize NaNs
+        withNewTempLocalDefs(targs) { (localDefs, cont1) =>
+          val argLocalDef = localDefs.head
+          def argToDouble = UnaryOp(UnaryOp.FloatToDouble, argLocalDef.newReplacement)
+          cont1 {
+            If(BinaryOp(BinaryOp.Double_!=, argToDouble, argToDouble),
+                IntLiteral(java.lang.Float.floatToIntBits(Float.NaN)),
+                wasmUnaryOp(WasmUnaryOp.I32ReinterpretF32, argLocalDef.toPreTransform))(
+                IntType).toPreTransform
+          }
+        } (cont)
+
+      case IntBitsToFloat =>
+        contTree(wasmUnaryOp(WasmUnaryOp.F32ReinterpretI32, targs.head))
+
+      // java.lang.Double
+
+      case DoubleToLongBits =>
+        // The Wasm I64ReinterpretF64 is the *raw* version; we need to normalize NaNs
+        withNewTempLocalDefs(targs) { (localDefs, cont1) =>
+          val argLocalDef = localDefs.head
+          cont1 {
+            If(BinaryOp(BinaryOp.Double_!=, argLocalDef.newReplacement, argLocalDef.newReplacement),
+                LongLiteral(java.lang.Double.doubleToLongBits(Double.NaN)),
+                wasmUnaryOp(WasmUnaryOp.I64ReinterpretF64, argLocalDef.toPreTransform))(
+                LongType).toPreTransform
+          }
+        } (cont)
+
+      case LongBitsToDouble =>
+        contTree(wasmUnaryOp(WasmUnaryOp.F64ReinterpretI64, targs.head))
+
+      // java.lang.Math
+
+      case MathAbsFloat =>
+        contTree(wasmUnaryOp(WasmUnaryOp.F32Abs, targs.head))
+      case MathAbsDouble =>
+        contTree(wasmUnaryOp(WasmUnaryOp.F64Abs, targs.head))
+      case MathCeil =>
+        contTree(wasmUnaryOp(WasmUnaryOp.F64Ceil, targs.head))
+      case MathFloor =>
+        contTree(wasmUnaryOp(WasmUnaryOp.F64Floor, targs.head))
+      case MathRint =>
+        contTree(wasmUnaryOp(WasmUnaryOp.F64Nearest, targs.head))
```

Comment:
I this correct for non-canonical NaNs?

> fnearest: If is a NaN, then return an element of nansN{z}

I do not understand the specification of nans_N, but IIUC this doesn't guarantee exact NaN propagation (but the JDK does).

Reply:
We never expose the bits of a NaN, so we don't have to care which NaN we produce.

Reply:
Hmm.. OK. I'm a bit concerned that im it might be too easy to forget at some point that we made this assumption, but maybe we just have to live with that.

Reply:
I think we've already made that assumption in many user-land implementations within the JDK. If we do allow observing NaN bit patterns at some point, we'll have to make a broader audit of all functions like that. I don't think we're introducing anything new here.